server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java

```diff
@@ -316,7 +318,211 @@ public RecordList scan(DataStoreScanRequest req) {
+                        while (!records.isEmpty()) {
+                            lastKey = Collections.min(records, (a, b) -> {
+                                var x = a.record.getKey();
+                                var y = b.record.getKey();
+                                return x.compareTo(y);
+                            }).record.getKey();
+                            for (var r : records) {
+                                if (r.record.getKey().equals(lastKey)) {
+                                    if (!r.record.isDeleted() && !keys.contains(r.record.getKey())) {
+                                        keys.add(r.record.getKey());
+                                    }
+                                    if (r.iterator.hasNext()) {
+                                        r.record = r.iterator.next();
+                                    } else {
+                                        r.record = null;
+                                    }
+                                }
+                            }
+
+                            records.removeIf(r -> r.record == null);
+                        }
+
+                        var ranges = new ArrayList<KeyRangeList.Range>();
+                        var batchSize = req.getRangeInfo().getBatchSize();
+                        var index = 0;
+                        while (index < keys.size()) {
+                            if (index + batchSize < keys.size()) {
+                                var startKey = keys.get(index);
+                                var start = (String) startKey.encode(req.isRawResult(), false);
+                                var startType = startKey.getColumnType().name();
+
+                                var endKey = keys.get(index + batchSize);
+                                var end = (String) endKey.encode(req.isRawResult(), false);
+                                var endType = endKey.getColumnType().name();
+                                ranges.add(KeyRangeList.Range.builder()
+                                        .start(start)
+                                        .startType(startType)
+                                        .startInclusive(true)
+                                        .end(end)
+                                        .endType(endType)
+                                        .endInclusive(false)
+                                        .size(batchSize)
+                                        .build());
+                                index += batchSize;
+                            } else {
+                                var startKey = keys.get(index);
+                                var start = (String) startKey.encode(req.isRawResult(), false);
+                                var startType = startKey.getColumnType().name();
+
+                                var endKey = keys.get(keys.size() - 1);
+                                var end = (String) endKey.encode(req.isRawResult(), false);
```

Comment:
The encoded key can not use raw result, the `rawResult` is for response record values, not keys.

Reply:
fixed, use false